# Chapter1. 신경망 복습

## 1.1 수학과 파이썬 복습

### 1.1.1 벡터와 행렬

* 벡터 : 크기와 방향을 가진 양
* 행렬 : 숫자가 2차원 형태로 늘어선 것

<img src="./images/1-1.png" width="300" height="300"/>
[1-1]

* 텐서 : 벡터와 행렬을 확장하여 숫자 집합을 N차원으로 표현

<img src="./images/1-2.png" width="300" height="300"/>
[1-2]

In [1]:
import numpy as np

In [2]:
x = np.array([1,2,3])

In [3]:
x.__class__

numpy.ndarray

In [4]:
x.shape

(3,)

In [5]:
x.ndim

1

In [6]:
W = np.array([[1,2,3],[4,5,6]])

In [7]:
W.shape

(2, 3)

In [8]:
W.ndim

2

### 1.1.2 행렬의 원소별 연산

In [9]:
W = np.array([[1,2,3],[4,5,6]])
X = np.array([[0,1,2],[3,4,5]])
W + X

array([[ 1,  3,  5],
       [ 7,  9, 11]])

In [10]:
W * X # 행렬 같은 자리별 곱셈이 이뤄짐

array([[ 0,  2,  6],
       [12, 20, 30]])

### 1.1.3 브로드캐스트

In [11]:
A = np.array([[1,2],[3,4]])
A * 10

array([[10, 20],
       [30, 40]])

In [12]:
A = np.array([[1,2],[3,4]])
b = np.array([10, 20])

In [13]:
A * b

array([[10, 40],
       [30, 80]])

### 1.1.4 벡터의 내적과 행렬의 곱

<img src="./images/1.1.png" width="200" height="200"/>
[1.1]

<img src="./images/1-5.png" width="300" height="300"/>
[1-5]

In [14]:
# 내적
a = np.array([1,2,3])
b = np.array([4,5,6])
np.dot(a,b)

32

In [15]:
# 행렬곱
A = np.array([[1,2],[3,4]])
B = np.array([[5,6],[7,8]])
np.matmul(A,B) # matrix mul

array([[19, 22],
       [43, 50]])

In [16]:
np.dot(A,B) # matrix mul

array([[19, 22],
       [43, 50]])

#### 잠깐! (dot 과 matmul의 차이)

* 2차원 이하의 행렬곱에서는 결과 동일
* 3차원 이상의 행렬끼리 곱할 때 다른 결과를 보여줌
* np.dot(a, b)는 a의 마지막 axis와 b의 마지막에서 두번째 axis끼리 곱한 결과를 반영하고, np.matmul(a, b)는 a의 마지막 두 axis 차원이 (n, k), b의 마지막 두 axis 차원이 (k, m) 일 때, 이 사이에서만 곱하여 (n, m) 차원이 되도록 반영


In [17]:
A = np.arange(2*3*4).reshape((2,3,4))
B = np.arange(2*3*4).reshape((2,4,3))

In [18]:
A

array([[[ 0,  1,  2,  3],
        [ 4,  5,  6,  7],
        [ 8,  9, 10, 11]],

       [[12, 13, 14, 15],
        [16, 17, 18, 19],
        [20, 21, 22, 23]]])

In [19]:
B

array([[[ 0,  1,  2],
        [ 3,  4,  5],
        [ 6,  7,  8],
        [ 9, 10, 11]],

       [[12, 13, 14],
        [15, 16, 17],
        [18, 19, 20],
        [21, 22, 23]]])

In [20]:
np.dot(A,B).shape

(2, 3, 2, 3)

In [21]:
np.matmul(A,B).shape

(2, 3, 3)

In [22]:
np.dot(A,B)

array([[[[  42,   48,   54],
         [ 114,  120,  126]],

        [[ 114,  136,  158],
         [ 378,  400,  422]],

        [[ 186,  224,  262],
         [ 642,  680,  718]]],


       [[[ 258,  312,  366],
         [ 906,  960, 1014]],

        [[ 330,  400,  470],
         [1170, 1240, 1310]],

        [[ 402,  488,  574],
         [1434, 1520, 1606]]]])

In [23]:
np.matmul(A,B)

array([[[  42,   48,   54],
        [ 114,  136,  158],
        [ 186,  224,  262]],

       [[ 906,  960, 1014],
        [1170, 1240, 1310],
        [1434, 1520, 1606]]])

### 1.1.5 행렬 형상 확인

<img src="./images/1-6.png" width="400" height="400"/>
[1-6]

## 1.2 신경망의 추론

### 1.2.1 신경망 추론 전체 그림

<img src="./images/1-7.png" width="300" height="300"/>
[1-7]

* 완전연결계층 : 인전합는 츠으이 모든 뉴런과 연결되어 있는 신경망

#### (수식) 입력 데이터 : $(x_1, x_2)$ 가중치 : $w_{11}, w_{12}$ 편향 : $b_1$

<img src="./images/1.2.png" width="150" height="150"/>
[1.2]

(예시)

<img src="./images/1.3.png" width="400" height="400"/>
[1.3]

<img src="./images/1.4.png" width="100" height="100"/>
[1.4]

#### (입력데이터가 1개)

<img src="./images/1-8.png" width="300" height="300"/>
[1-8]

#### (1개의 입력데이터가 아니라 N개의 입력데이터라면)

<img src="./images/1-9.png" width="300" height="300"/>
[1-9]

In [24]:
## 완전연결계층 미니배치

In [25]:
import numpy as np
W1 = np.random.randn(2,4)
b1 = np.random.randn(4)
x = np.random.randn(10, 2)
h = np.matmul(x, W1) + b1

In [26]:
h

array([[ 0.17247735, -0.87835075,  0.11987837,  0.53045699],
       [-0.90193738, -0.90952458,  0.11748406,  1.03878783],
       [-0.58387496,  0.46201707,  0.39971208,  0.52296337],
       [ 1.61608433,  2.44124176,  0.80042679, -1.03155437],
       [ 2.29751267,  1.04216265,  0.50874278, -0.9734504 ],
       [ 1.95074845,  2.65066869,  0.84244364, -1.24345136],
       [-1.98853691, -0.68203101,  0.16858874,  1.48342002],
       [ 1.93341007,  1.72898318,  0.65204453, -0.98820781],
       [ 1.84447346,  2.7484224 ,  0.86304461, -1.22021244],
       [-1.43411328, -0.32176902,  0.24094744,  1.12880906]])

#### (완전연결계층에 의한 변환 : 선형 변환 / 비선형 효과 부여 : 활성화 함수)

* 예시 
    * 시그모이드 함수

<img src="./images/1.5.png" width="200" height="200"/>
[1.5]

<img src="./images/1-10.png" width="300" height="300"/>
[1-10]

In [27]:
## 시그모이드 함수 구현

In [28]:
def sigmoid(x):
    return 1 / (1 + np.exp(-x))

In [29]:
a = sigmoid(h)

In [30]:
a

array([[0.54301276, 0.29351966, 0.52993375, 0.62958969],
       [0.28865253, 0.28709713, 0.52933728, 0.73861605],
       [0.35804145, 0.61349257, 0.59861848, 0.62784044],
       [0.8342544 , 0.91991861, 0.69006577, 0.26278287],
       [0.90867083, 0.73926707, 0.62451171, 0.2741933 ],
       [0.87552823, 0.93405219, 0.69897962, 0.2238358 ],
       [0.12041174, 0.33580815, 0.54204764, 0.8150886 ],
       [0.87362638, 0.84928231, 0.65747105, 0.27126621],
       [0.86347692, 0.93982419, 0.70329637, 0.22789907],
       [0.1924586 , 0.42024468, 0.55994712, 0.75561905]])

### 1.2.2 계층으로 클래스화 및 순전파 구현

* (구현 규칙)
    * 모든 계층은 forward(), backward() 메서드를 가진다
        * forward : 순전파 
        * backward : 역전파
    * 모든 계층은 인스턴스 변수인 params, grads를 가진다
        * params : 가중치와 편향 등 매개변수 담는 리스트
        * grads : params에 저장된 각 매개변수에 대응하여 해당 매개변수의 기울기를 보관하는 리스트

#### (이 절에서는 순전파만 구현)

In [31]:
# Sigmoid 계층 구현

In [32]:
import numpy as np

In [33]:
class Sigmoid:
    def __init__(self):
        self.params = []
    
    def forward(self, x):
        return 1 / (1 + np.exp(-x))

In [34]:
class Affine:
    def __init__(self, W, b):
        self.params = [W, b]
        
    def forward(self, x):
        W, b = self.params ### 리스트는 사이즈가 같을 때 각각의 번지의 값을 바로 넣을 수 있다
        out = np.matmul(x,W) + b
        return out

<img src="./images/1-11.png" width="400" height="400"/>
[1-11]

In [35]:
class TwoLayerNet:
    def __init__(self, input_size, hidden_size, output_size):
        I, H, O = input_size, hidden_size, output_size
        
        # 가중치와 편향 초기화
        W1 = np.random.randn(I,H)
        b1 = np.random.randn(H)
        W2 = np.random.randn(H,O)
        b2 = np.random.randn(O)
        
        # 계층 생성
        self.layers = [Affine(W1, b1), Sigmoid(), Affine(W2, b2)]
        
        # 모든 가중치를 리스트에 모은다
        self.params = []
        for layer in self.layers:
            self.params += layer.params ## extend와 같은 듯
            
    def predict(self, x):
        for layer in self.layers:
            x = layer.forward(x)

        return x

In [36]:
### 리슽와 +

In [37]:
a = ['A', 'B']
a += ['C', 'D']
a

['A', 'B', 'C', 'D']

In [38]:
x = np.random.randn(10,2)
model = TwoLayerNet(2,4,3)
s = model.predict(x)

In [39]:
s

array([[ 0.18665613, -2.13251884,  3.32341811],
       [-0.08278338, -1.60286795,  1.80344187],
       [ 0.17543506, -2.37140632,  3.69759982],
       [ 0.00885726, -1.98306677,  2.60798545],
       [-0.08559552, -1.7223268 ,  1.98294111],
       [-0.02795212, -1.73857919,  2.13984152],
       [ 0.12572249, -1.95670015,  2.89112643],
       [-0.10944917, -1.64132461,  1.82148649],
       [ 0.06577315, -1.67970484,  2.23979903],
       [ 0.07257299, -2.22791235,  3.16932172]])

## 1.3 신경망의 학습

### 1.3.1 손실 함수

* 손실 : 학습데이터와 신경망이 예측한 결과를 비교하여 예측이 얼마나 나쁜가를 산출한 스칼라 값
* 손실 함수
    * 다중 클래스 분류 : 교차 엔트로피 오차 등 이용

<img src="./images/1-12.png" width="500" height="500"/>
[1-12]

##### (소프트맥스 함수)

<img src="./images/1.6.png" width="100" height="100"/>
[1.6]

#### (교차 엔트로피 오차)

<img src="./images/1.7.png" width="100" height="100"/>
[1.7]

#### (미니 배치를 고려한 교차 엔트로피 오차)

<img src="./images/1.8.png" width="150" height="150"/>
[1.8]


#### (Softmax와 Loss 계층을 합친다 : 역전파 계산이 쉬워짐)

<img src="./images/1-13.png" width="500" height="500"/>
[1-13]

### 1.3.2 미분과 기울기

* 신경망 학습의 목표 : 손실을 최소화하는 매개변수를 찾기

$\displaystyle L $ : 스칼라  $\displaystyle \textbf{x} $ : 벡터 <br>
$\displaystyle L = f(\textbf{x})$ 이고 $ \textbf{x} = (x_1, x_2,...,x_n)$ 일 때,<br>
$x_i$에 대한 $L$의 미분 : $\displaystyle \frac {\partial L}{\partial x_i}$이고<br>

$\displaystyle \frac{\partial L}{\partial \textbf{x} } =  \left
(\frac{\partial L}{\partial x_1},\frac{\partial L}{\partial x_2},...,\frac{\partial L}{\partial x_n} \right )$


$\displaystyle \textbf{W}$ : m x n 행렬일 때, <br>
$\displaystyle L = g(\textbf{W})$에 대해 <br>

<img src="./images/1.10.png" width="200" height="200"/>


### 1.3.3 연쇄법칙

* 오차역전파법
    * 연쇄 법칙 : 합성함수에 대한 미분 법칙
    * 신경망은 여러 함수가 연결된 것이라고 볼 수 있으므로 오차역전파법은 그 여러 함수에 대해 연쇄 법칙을 효율적으로 적용하여 기울기를 산출

<img src="./images/1.11.png" width="100" height="100"/>
[1.11]

### 1.3.4 계산 그래프

<img src="./images/1-15.png" width="200" height="200"/>
[1-15]

<img src="./images/1-16.png" width="500" height="500"/>
[1-16]

<img src="./images/1-17.png" width="500" height="500"/>
[1-17]

<img src="./images/1-18.png" width="400" height="400"/>
[1-18]

#### (곱셈 노드)

<img src="./images/1-19.png" width="500" height="500"/>
[1-19]

#### (분기 노드)

<img src="./images/1-20.png" width="500" height="500"/>
[1-20]

#### (Repeat 노드) 분기 노드를 N개 노드로 일반화

<img src="./images/1-21.png" width="500" height="500"/>
[1-21]

In [58]:
# 구현예시

In [59]:
import numpy as np

In [60]:
D, N = 8, 7
x = np.random.randn(1,D)
y = np.repeat(x, N, axis= 0) ## 복제함

In [61]:
x

array([[ 0.52392298,  0.86899412, -0.72400921, -0.4397237 ,  0.53191192,
         1.57778249, -1.96893978, -1.85600151]])

In [62]:
y

array([[ 0.52392298,  0.86899412, -0.72400921, -0.4397237 ,  0.53191192,
         1.57778249, -1.96893978, -1.85600151],
       [ 0.52392298,  0.86899412, -0.72400921, -0.4397237 ,  0.53191192,
         1.57778249, -1.96893978, -1.85600151],
       [ 0.52392298,  0.86899412, -0.72400921, -0.4397237 ,  0.53191192,
         1.57778249, -1.96893978, -1.85600151],
       [ 0.52392298,  0.86899412, -0.72400921, -0.4397237 ,  0.53191192,
         1.57778249, -1.96893978, -1.85600151],
       [ 0.52392298,  0.86899412, -0.72400921, -0.4397237 ,  0.53191192,
         1.57778249, -1.96893978, -1.85600151],
       [ 0.52392298,  0.86899412, -0.72400921, -0.4397237 ,  0.53191192,
         1.57778249, -1.96893978, -1.85600151],
       [ 0.52392298,  0.86899412, -0.72400921, -0.4397237 ,  0.53191192,
         1.57778249, -1.96893978, -1.85600151]])

In [75]:
dy = np.random.randn(N, D)
dx = np.sum(dy, axis=0, keepdims=True)

In [76]:
dy

array([[-1.21885848,  0.50218897, -1.33640858,  0.36324557, -0.13566862,
        -0.03979735,  0.83631607,  0.04923945],
       [ 0.75738157, -1.34040954, -0.32495492, -2.80306581,  0.21816735,
         0.67977637, -0.26269876, -0.99243605],
       [ 0.88729455,  0.68536921,  1.54074605, -0.03126707,  1.14915251,
         0.24551681,  1.39405281,  0.59222012],
       [-0.07617761,  0.23836602, -1.53341768,  1.31130544,  1.65766609,
         1.15518555, -0.20843097, -0.16497338],
       [ 0.37053573,  1.03949722, -0.88759078,  0.82490725, -1.25967134,
        -1.35755654,  0.7307986 , -1.59173231],
       [-1.03082256,  0.13364008, -0.65590994,  0.00929562, -2.35531691,
        -0.26283458,  0.77995183,  0.35940215],
       [ 0.64897791,  0.96672238, -0.39757048, -0.85317021, -0.7082987 ,
        -0.76697872, -1.1113733 , -0.80958939]])

In [77]:
dx

array([[ 0.33833111,  2.22537434, -3.59510634, -1.1787492 , -1.43396961,
        -0.34668846,  2.15861628, -2.55786941]])

#### (Sum 노드)

<img src="./images/1-22.png" width="500" height="500"/>
[1-22]

In [78]:
# 구현 예시

In [40]:
D, N = 8 ,7

In [41]:
x = np.random.randn(N,D)

In [42]:
x

array([[ 0.85759692, -0.91609426,  1.37265356, -0.60454383, -1.68572155,
         1.88539746,  0.92151078,  0.89574257],
       [ 1.91157955,  2.11727639,  0.67781817, -1.41412163,  1.05064701,
        -0.94469163, -0.22954469, -0.5168888 ],
       [-0.50923283, -0.66314813,  1.19723616,  0.33257567, -2.37086622,
        -0.17818606,  0.48526746,  2.18544369],
       [ 0.9526694 , -0.9870124 ,  0.57196106, -0.73263299,  0.41571631,
        -0.55240609,  0.16944017, -0.50101077],
       [-0.45633677,  0.27893291,  1.19450568,  0.02020904,  0.6664018 ,
        -0.35011964, -1.08317795,  1.57056888],
       [ 0.98759625,  0.41029056,  0.99652058, -0.91815167, -0.8671486 ,
        -0.74612137,  0.41157851, -0.36271063],
       [-0.53685705,  1.90766043,  0.45955159, -0.10691491,  0.77148695,
         0.21009337, -0.01935291,  1.09668555]])

In [43]:
y = np.sum(x, axis=0, keepdims = True)

In [44]:
y

array([[ 3.20701546,  2.1479055 ,  6.47024679, -3.42358032, -2.0194843 ,
        -0.67603395,  0.65572136,  4.36783051]])

In [79]:
dy = np.random.randn(1,D)
dx = np.repeat(dy,N,axis=0)

#### (MatMul 노드)

해당 내용은 1권에서 작성한 [여기]("https://github.com/Dongyeoni/Deep-Learning-from-Scratch/blob/main/5%EC%9E%A5.ipynb")의 5.6절을 참고하면 좋을 것 같다

$\displaystyle \textbf{x} : $1 X D,
$\textbf{W}$ : D X H, 
$\textbf{y}$ : 1 X H <br>
$\displaystyle \textbf{y = xW}$ 일 때

<img src="./images/1-23.png" width="300" height="300"/>
[1-23]

$\displaystyle \textbf{x}$의 i 번째 원소에 대한 미분

<img src="./images/1.12.png" width="125" height="125"/>
[1.12]

$\displaystyle \textbf{y} = \textbf{xW}$ 이므로 각 $i$ 에 대해서 <br><br>
$\displaystyle \frac {\partial y_j} {\partial x_i} = W_{ij}$ 가 성립한다
정리하면 아래와 같다

<img src="./images/1.13.png" width="200" height="200"/>
[1.13]

다시 생각해보자.
$\displaystyle \textbf{x} : $1 X D,
$\textbf{W}$ : D X H, 
$\textbf{y}$ : 1 X H 이므로<br>
$j$ 값이 1부터 H까지라고 할 수 있다

$\displaystyle \frac {\partial L} {\partial y_1} W_{i1} + \frac {\partial L} {\partial y_2} W_{i2}+ ... \frac {\partial L} {\partial y_H} W_{iH} $ <br>

$\displaystyle  
\frac {\partial L} {\partial x_i} = 
\begin{pmatrix}
 \frac {\partial L}{\partial y_1}
 & ...
 &  \frac {\partial L}{\partial y_H}
\end{pmatrix}
\begin{pmatrix}
W_{i1} \\ ... \\ W_{iH} \\
\end{pmatrix} $ <br>
$\displaystyle  = \frac {\partial L} {\partial y} 
\begin{pmatrix}
W_{i1} & ... & W_{iH} 
\end{pmatrix}^T
$ 이다.


해석하자면 $\displaystyle L$을 $\displaystyle x_i$로 편미분한 결과는
$\displaystyle L$을 $\displaystyle y$로 편미분한 결과와 $\displaystyle W$의 $i$번째 행을 transpose한 후 내적한 값과 같다. <br>
식을 정리하면 아래와 같다.(헤깔리지 않게 대문자 W를 대신 소문자 w를 쓸 걸 그랬다...)

<img src="./images/1.14.png" width="125" height="125"/>
[1.14]

### (형상 확인하기)

<img src="./images/1-24.png" width="300" height="300"/>
[1-24]